Fixes testing network APIs + adds 3 new APIs (btc.com, blockchair.com and blockstream.info)

bjarnemagnussen opened the pull request:
*Due to time-out errors adding more network API's makes it even more difficult to pass all the tests. This problem still persists with this PR and may make it worse due to the three new APIs!*

## Fix for testing balance > 0
This PR fixes the testing error from `test_get_balance_main_used` inside `BitpayAPI`, `BlockchainAPI` and `SmartbitAPI`. 

The reason was that the address being checked was used up and has a balance of 0, while the tests are asserting that the balance must be greater than 0.

It has been fixed by referring to Counterparty's burn-of-proof address `1CounterpartyXXXXXXXXXXXXXXXUWLpVr` and should therefore always have a balance greater than 0.

## Adding network APIs

The PR further adds three new network APIs: 
* Btc.com, 
* Blockstream.info, and 
* Blockchair.com. 

Blockchair was the most straight-forward to implement as it includes a "context" with each response that is very useful in case the immediate response does not include information that we need (e.g. confirmations, scriptPubKey, etc.). However for the first two APIs there were minor issues that I want to mention and eventually discuss:

1. For the new `BTCcomAPI` and `BlockstreamAPI`: 
When unspents are returned the scriptPubKey of the address is *not* included with the response. We must therefore calculate it from the address using our helper function `address_to_scriptpubkey` found inside `bit.transaction`.
2. For the new BlockstreamAPI: 
When unspents are returned it will include the block number the transaction was confirmed in, but neither include the number of confirmations nor the current block height. We must therefore first poll the current block height to then be able to calculate the confirmations for each unspent transaction returned.
3. For the new `BlockstreamAPI`:
Returned unspents are not sorted by any metric and appear to be random. To compare them directly with unspents returned from other APIs the returned unspents are immediately sorted by their confirmation number.

## Future development

If this PR has been checked and should be approved, then the next step will be to add a features-supported functionality, allowing `NetworkAPI` to poll the APIs with required features supported depending on the use-case, e.g. for bech32 addresses.

bjarnemagnussen commented:
Uff, seems that there is a bot using our testing private key(s) to steal them [immediately](https://www.blockchain.com/btctest/address/mtrNwJxS1VyHYn3qBY1Qfsm3K3kh1mGRMS)... Maybe it is possible to set private keys in environment variables and configure them privately in Travis CI?

teran-mckinney commented:
Wow, nice work!

I do think that Travis can hold secrets. I wonder if someone just has those private keys and if we change them that things will be fine for a while, though. No idea why anyone wants to steal testnet coins.

Will need to get back to this later to review it.

bjarnemagnussen commented:
Yeah, having a bot steal testnet bitcoins is ridiculous... :)

ofek commented:
Please email me and tell me what to store

bjarnemagnussen commented:
I have fixed one typo in `BTCcomAPI` and added a longer timeout when testing the services.

If we could add the new APIs with this PR, then I can add code that ultimately allows for Bech32 addresses to be polled :)

teran-mckinney commented:
With three new APIs in one PR, will one more hurt? :)

I'll want to test this over Tor to see if it behaves properly still.

ofek commented:
Rebase?

teran-mckinney commented:
I'm sorry for the wait! Completely forgot to test this.

bjarnemagnussen commented:
Great! I will take a look at it and rebase in the next days.

I think we really should clean up the APIs as Insight by BitPay seems to really have bitten the dust and the project page now also leads to bitcore.io (https://github.com/bitpay/insight/). I may additionally add bitcoire.io, as they seem to have an up and working API.

teran-mckinney commented:
If it helps, we added Bitcore to bitcash: https://github.com/sporestack/bitcash/blob/master/bitcash/network/services.py#L252

bjarnemagnussen commented:
Just a point I would like to discuss:
Maintaining all of the APIs can become a burden, and it is an annoyance that the tests report errors just because a single API raises an error.

More specifically a warning is currently only logged in testing for the individual APIs if they fail due to `ConnectionError`, etc, as this can often happen. However if they fail inside the test of the `NetworkAPI` class, it is shown as an error in testing.
In the implementation `NetworkAPI` skips over any API that raise ignored errors, and only raises an error if _all_ fail.

I was therefore thinking of changing the test of `NetworkAPI` slightly so that it will try to poll all the API's, but only log a warning if _all_ of them fail? Does it make sense?

ofek commented:
Sounds fine 👍

teran-mckinney commented:
I tend to think it's nice to know what errors were had on a given API. Otherwise, it's almost impossible to debug. So a logging.Warning per NetworkError, of course but only raising an exception if all fail.

bjarnemagnussen commented:
@teran-mckinney Yes, just to clarify:
Currently inside the tests for each of the APIs a warning is logged for e.g. `ConnectionError` (part of the "ignored errors"). 

However, when testing the `NetworkAPI` class each API is again called but this time no errors are ignored. Instead the whole test fails if just a single API has e.g. a `ConnectionError`. This is where I think an error should only be raised if _all_ APIs fail inside, but also expecting errors of single APIs to be logged in their own tests.

teran-mckinney commented:
Ah, gotcha. Thank you for explaining. I agree with you.

bjarnemagnussen commented:
I have pushed all the changes with a new commit. After manually restarting Travis testing it works, however sending is still a problem.

I think the environment variables `WALLET_FORMAT_SEND_TEST` and `WALLET_FORMAT_COMPRESSED_SEND_TEST` are not available and therefore _new_ private keys are instead generated that don't have any unspents. This may be because the variables are protected and not available to PRs from another repository (mine).

ofek commented:
Ah yes that is it. Make branch here?

bjarnemagnussen commented:
Changed branch to allow Travis to hopefully work. Continued with #112
